From 764a528cae30309ff73fc2b15fabebe08c95b823 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?S=C3=B6ren=20Tempel?= <soeren+git@soeren-tempel.net>
Date: Wed, 2 Sep 2026 05:26:51 +0200
Subject: [PATCH] modprobe: Check for ELF header to determine if module is
 compressed

Without this patch, BusyBox calls finit_module(2) twice. Once with, and
once w/o MODULE_INIT_COMPRESSED_FILE. On Alpine, the first call fails
because the kernel is configured without kernel-space module
decompression (i.e., CONFIG_MODULE_DECOMPRESS is not set). The second
call, which attempts to load a compressed module, without indicating
that it is compressed, then causes the following warning to be emitted:

        Invalid ELF header magic: != ELF

This is fixed in this patch by checking if the file contains an ELF
header to check if it is compressed. Only if no ELF header was found,
the MODULE_INIT_COMPRESSED_FILE is used. A similar check is already
done in bb_init_module() anyhow, so also do this check before calling
finit_module().

Not implemented for modprobe-small as we don't use that on Alpine.

See also:

* We used to carry such a patch before the upgrade to Busybox 1.37.0,
  which rewrote the loading logic and added the opportunistic loop.
  See <https://git.alpinelinux.org/aports/commit/?id=531e69840c5ae44>.
  This patch was written by Qu Wenruo <wqu@suse.com>:
    * http://lists.busybox.net/pipermail/busybox/2021-January/088398.html
    * http://lists.busybox.net/pipermail/busybox/2021-January/088399.html
* A patch which open the file twice to check for the header exists.
  See <https://lists.busybox.net/pipermail/busybox/2024-July/090802.html>.
  This patch was written by Yao Zi <me@ziyao.cc>.
* https://gitlab.alpinelinux.org/alpine/aports/-/work_items/13427
* https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/30322

Co-Authored-By: Qu Wenruo <wqu@suse.com>
Co-Authored-By: Yao Zi <me@ziyao.cc>
---
 modutils/modutils.c | 38 +++++++++++++++++++++++++++++---------
 1 file changed, 29 insertions(+), 9 deletions(-)

diff --git a/modutils/modutils.c b/modutils/modutils.c
index c57564023..afcc82ce9 100644
--- a/modutils/modutils.c
+++ b/modutils/modutils.c
@@ -177,6 +177,10 @@ char* FAST_FUNC parse_cmdline_module_options(char **argv, int quote_spaces)
 }
 #endif
 
+static inline int is_elf_header(uint32_t word) {
+	return word == SWAP_BE32(0x7f454c46);
+}
+
 #if ENABLE_FEATURE_INSMOD_TRY_MMAP
 void* FAST_FUNC try_to_mmap_module(const char *filename, size_t *image_size_p)
 {
@@ -197,7 +201,7 @@ void* FAST_FUNC try_to_mmap_module(const char *filename, size_t *image_size_p)
 		image = mmap_read(fd, image_size);
 		if (image == MAP_FAILED) {
 			image = NULL;
-		} else if (*(uint32_t*)image != SWAP_BE32(0x7f454C46)) {
+		} else if (!is_elf_header(*(uint32_t*)image) {
 			/* No ELF signature. Compressed module? */
 			munmap(image, image_size);
 			image = NULL;
@@ -211,6 +215,22 @@ void* FAST_FUNC try_to_mmap_module(const char *filename, size_t *image_size_p)
 }
 #endif
 
+#ifdef __NR_finit_module
+static int has_elf_header(int fd)
+{
+	uint32_t signature;
+	int ret;
+
+	/* use pread(3) here to not change the file offset. in theory,
+	 * a {full,safe}_read() variant for pread would be useful here. */
+	ret = pread(fd, &signature, sizeof(signature), 0);
+	if (ret < sizeof(signature))
+		return 0;
+
+	return is_elf_header(signature);
+}
+#endif
+
 /* Return:
  * 0 on success,
  * -errno on open/read error,
@@ -237,19 +257,19 @@ int FAST_FUNC bb_init_module(const char *filename, const char *options)
 	 * to only allow loading of modules off of secure storage (like a read-
 	 * only rootfs) which needs the finit_module call.  If it fails, we fall
 	 * back to normal module loading to support compressed modules.
+	 *
+	 * Note that finit_module()/init_module() only accept ELF image, do some
+	 * basic check before calling finit_module() to prevent the kernel from
+	 * complaining about "invalid ELF header magic" in the kernel ring buffer.
+	 *
+	 * See https://gitlab.alpinelinux.org/alpine/aports/-/work_items/13427
 	 */
 # ifdef __NR_finit_module
 	{
 		int fd = open(filename, O_RDONLY | O_CLOEXEC);
 		if (fd >= 0) {
-			int flags = is_suffixed_with(filename, ".ko") ? 0 : MODULE_INIT_COMPRESSED_FILE;
-			for (;;) {
-				rc = finit_module(fd, options, flags);
-				if (rc == 0 || flags == 0)
-					break;
-				/* Loading non-.ko named uncompressed module? Not likely, but let's try it */
-				flags = 0;
-			}
+			int flags = has_elf_header(fd) ? 0 : MODULE_INIT_COMPRESSED_FILE;
+			rc = finit_module(fd, options, flags);
 			close(fd);
 			if (rc == 0)
 				return rc;
